lib/rules/no-component-lifecycle-hooks.js

```diff
@@ -0,0 +1,55 @@
+'use strict';
+
+const emberUtils = require('../utils/ember');
+
+const { isComponentLifecycleHookName } = emberUtils;
+
+const ERROR_MESSAGE_NO_COMPONENT_LIFECYCLE_HOOKS =
+  'Do not use classic ember components lifecycle hooks. Prefer using "@ember/render-modifiers" or custom functional modifiers.';
+
+module.exports = {
+  ERROR_MESSAGE_NO_COMPONENT_LIFECYCLE_HOOKS,
+  meta: {
+    docs: {
+      description:
+        'Prevents usage of "classic" ember component lifecycle hooks. Render modifiers or custom functional modifiers should be used instead.',
+      category: 'Ember Octane',
+      recommended: false,
+    },
+    fixable: null, // or "code" or "whitespace"
+  },
+
+  create(context) {
+    let isAnEmberComponent = false;
+
+    const isAClassicLifecycleHook = keyName => {
+      return isComponentLifecycleHookName(keyName) && isAnEmberComponent;
+    };
+
+    const report = (context, node) => {
+      context.report(node, ERROR_MESSAGE_NO_COMPONENT_LIFECYCLE_HOOKS);
+    };
+
+    return {
+      ClassDeclaration(node) {
```

Comment:
You need to reset isAnEmberComponent in `MemberExpression:exit` and `ClassDeclaration:exit` in case there are multiple components inside a file.

Reply:
do you mean cases such like this one:

```js
import Component from '@ember/component';

export const Component2 = Component.extend({
  didDestroyElement() {},
});

export const Component1 = Component.extend({
  willDestroy() {},
);
```
? If so the tests for such case are still passing. Am I missing something?

Reply:
Basically I meant that we should know when we are no longer inside a component. In this following example, we are initially in a component, but then once we leave it, but then we might be inside a random class or object after, where we shouldn't be checking for hooks anymore.

```
export default class MyComponent extends Component {
  // ...
}

const someRandomClassOrObject = { didDestroyElement() { } };
```

Reply:
gotcha! Will add a test case for that, thanks!